Gravatar/MruImageCache.cs

```diff
@@ -0,0 +1,171 @@
+                throw new ArgumentException(nameof(imageFileName));
+            }
+
+            _entryByFileName.TryRemove(imageFileName, out _);
+            return _inner.DeleteImageAsync(imageFileName);
+        }
+
+        Image IImageCache.GetImage(string imageFileName)
+        {
+            if (string.IsNullOrWhiteSpace(imageFileName))
+            {
+                throw new ArgumentException(nameof(imageFileName));
+            }
+
+            if (_entryByFileName.TryGetValue(imageFileName, out var entry))
+            {
+                entry.LastAccesedAt = DateTime.UtcNow;
+                return entry.Image;
+            }
+
+            var image = _inner.GetImage(imageFileName);
+
+            if (image != null)
+            {
+                UpdateEntry(imageFileName, image);
+            }
+
+            return image;
+        }
+
+        async Task<Image> IImageCache.GetImageAsync(string imageFileName)
+        {
+            if (string.IsNullOrWhiteSpace(imageFileName))
+            {
+                throw new ArgumentException(nameof(imageFileName));
+            }
+
+            if (_entryByFileName.TryGetValue(imageFileName, out var entry))
+            {
+                entry.LastAccesedAt = DateTime.UtcNow;
+                return entry.Image;
+            }
+
+            var image = await _inner.GetImageAsync(imageFileName);
+
+            if (image != null)
+            {
+                UpdateEntry(imageFileName, image);
+            }
+
+            return image;
+        }
+
+        private void UpdateEntry([NotNull] string imageFileName, [NotNull] Image image)
+        {
+            if (string.IsNullOrWhiteSpace(imageFileName))
+            {
+                throw new ArgumentException(nameof(imageFileName));
+            }
+
+            if (image == null)
+            {
+                throw new ArgumentNullException(nameof(image));
+            }
+
+            _entryByFileName[imageFileName] = new Entry(imageFileName, image);
+
+            if (_entryByFileName.Count > _cleanAtSize)
+            {
+                RemoveOldEntries();
+            }
+
+            void RemoveOldEntries()
+            {
+                // Build list of all entries, sorted by their last access times
+                var sortedEntries = _entryByFileName.Values.OrderBy(e => e.LastAccesedAt).ToList();
+
+                // The first items in that sorted list are the oldest ones.
+                // Take as many as we need to remove, and remove them from the dictionary.
+                foreach (var entry in sortedEntries.Take(_entryByFileName.Count - _cleanToSize))
+                {
+                    _entryByFileName.TryRemove(entry.FileName, out _);
+                }
+            }
+        }
+
+        private sealed class Entry
+        {
+            [NotNull] public string FileName { get; }
+            [NotNull] public Image Image { get; }
+            public DateTime LastAccesedAt { get; set; }
```

Comment:
❕ Use `DateTimeOffset` instead of `DateTime` except where prevented by constraints outside the control of the application.

Reply:
For public dates that might leave the machine, that makes sense. In this case it doesn't make a difference, does it? It uses UTC to protect against the two times a year the clocks change. Worst case, the cache is rebuilt or order of item expiry is unfair.

Reply:
It's not a problem until it is. `DateTimeOffset` is a lightweight type that can be used uniformly to avoid the possibility of certain classes of problems. It works nearly identically to `DateTime` so it's painless to use instead of it.

Either way, this property will not be necessary after the switch to `LinkedList`.